src/imageops/mod.rs

```diff
@@ -190,6 +192,76 @@ where
     }
 }
 
+/// Fill the image with a linear vertical gradient
+/// 
+/// This function assumes a linear color space.
+/// 
+/// # Examples
+/// ```no_run
+/// use image::{Rgba, RgbaImage, Pixel};
+/// 
+/// fn main() {
+///     let mut img = RgbaImage::new(100, 100);
+///     let start = Rgba::from_slice(&[0, 128, 0, 0]);
+///     let end = Rgba::from_slice(&[255, 255, 255, 255]);
+/// 
+///     image::imageops::vertical_gradient(&mut img, start, end);
+///     img.save("vertical_gradient.png").unwrap();
+/// }
+pub fn vertical_gradient<S, P, I>(img: &mut I, start: &P, stop: &P)
+where
+    I: GenericImage<Pixel = P>,
+    P: Pixel<Subpixel = S> + 'static,
+    S: Primitive + Lerp + 'static
+{
+    for y in 0..img.height() {
+        let pixel = start.map2(stop, |a, b| {
+            let y = <S::SignedLarger as NumCast>::from(y).unwrap();
+            let height = <S::SignedLarger as NumCast>::from(img.height()).unwrap();
```

Comment:
This code has `i < n` however the inner impl of `lerp` is computed as if under the assumption `i <= n`. In particular, the upper endpoint is never reached. Since the inner block can only be reached for a non-empty image it seems possible to use `img.height() - 1` instead to avoid this.

Reply:
Nice catch. Fixed and added a test to verify the boundaries.